From: Armin Novak <armin.novak@thincast.com>
Date: Wed, 9 Jun 2021 13:05:33 +0200
Subject: Added WINPR_ASSERT

Added a WinPR macro that can be used for WinPR and FreeRDP to
replace standard C assert.
It is more verbose and logs issues to WLog which makes it easier
for us to trace problems in the field.

(cherry picked from commit b435e6db9b21bfe0a456a38f086f51151d3c20ad)
(cherry picked from commit 76660aab19020f45d398fe42100351960ad40bb2)
---
 winpr/include/winpr/assert.h | 49 ++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 49 insertions(+)
 create mode 100644 winpr/include/winpr/assert.h

diff --git a/winpr/include/winpr/assert.h b/winpr/include/winpr/assert.h
new file mode 100644
index 0000000..724f82b
--- /dev/null
+++ b/winpr/include/winpr/assert.h
@@ -0,0 +1,49 @@
+/**
+ * WinPR: Windows Portable Runtime
+ * Runtime ASSERT macros
+ *
+ * Copyright 2021 Armin Novak <armin.novak@thincast.com>
+ * Copyright 2021 Thincast Technologies GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef WINPR_ASSERT_H
+#define WINPR_ASSERT_H
+
+#include <stdlib.h>
+#include <winpr/winpr.h>
+#include <winpr/wtypes.h>
+#include <winpr/wlog.h>
+
+#define WINPR_ASSERT(cond)                                                                   \
+	do                                                                                       \
+	{                                                                                        \
+		if (!(cond))                                                                         \
+		{                                                                                    \
+			WLog_FATAL("com.freerdp.winpr.assert", "%s [%s:%s:%" PRIuz "]", #cond, __FILE__, \
+			           __FUNCTION__, __LINE__);                                              \
+			abort();                                                                         \
+		}                                                                                    \
+	} while (0)
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif /* WINPR_ERROR_H */
